gdk/x11/gdkglcontext-x11-eglx.c

```diff
@@ -410,6 +410,108 @@ gdk_x11_gl_context_end_frame (GdkGLContext   *context,
     eglSwapBuffers (edpy, esurface);
 }
 
+static gboolean
+gdk_x11_gl_context_texture_from_surface (GdkGLContext    *context,
+                                         cairo_surface_t *surface,
+                                         cairo_region_t  *region)
+{
+  GdkWindow *window = gdk_gl_context_get_window (context);
+  GdkDisplay *display = gdk_window_get_display (window);
+  GdkX11Display *display_x11 = GDK_X11_DISPLAY (display);
+  EGLDisplay edpy = gdk_x11_display_get_egl_display (display);
+  EGLImageKHR img = EGL_NO_IMAGE_KHR;
+  const EGLint attribs[] = { EGL_IMAGE_PRESERVED_KHR, EGL_TRUE, EGL_NONE };
+  double device_x_offset, device_y_offset;
+  cairo_rectangle_int_t rect;
+  int n_rects, i;
+  int unscaled_window_height;
+  int window_scale;
+  unsigned int texture_id;
+  double sx, sy;
+  float uscale, vscale;
+  GdkTexturedQuad *quads;
+
+  if (!display_x11->has_image_pixmap ||
+      cairo_surface_get_type (surface) != CAIRO_SURFACE_TYPE_XLIB)
+    return FALSE;
+
+  window_scale = gdk_window_get_scale_factor (window);
+  gdk_window_get_unscaled_size (window, NULL, &unscaled_window_height);
+
+  sx = sy = 1;
+  cairo_surface_get_device_scale (window->current_paint.surface, &sx, &sy);
+  cairo_surface_get_device_offset (surface, &device_x_offset, &device_y_offset);
+
+  glGenTextures (1, &texture_id);
+  glBindTexture (GL_TEXTURE_2D, texture_id);
+
+  glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+  glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+  img = eglCreateImageKHR (edpy, EGL_NO_CONTEXT, EGL_NATIVE_PIXMAP_KHR,
+                           (EGLClientBuffer)cairo_xlib_surface_get_drawable (surface),
+                           attribs);
+
+  if (img == EGL_NO_IMAGE_KHR)
```

Comment:
You're leaking the GL texture, here.

You should create the EGL image object before that.